File: dawn_go/tango.go
package main

import "fmt"

type auxTree struct {
	left  *auxTree
	key   int
	right *auxTree
}

func newAuxTree(key int) *auxTree {
	return &auxTree{key: key}
}

func (t *auxTree) insert(key int) {
	if key < t.key {
		if t.left == nil {
			t.left = newAuxTree(key)
		} else {
			t.left.insert(key)
		}
	} else {
		if t.right == nil {
			t.right = newAuxTree(key)
		} else {
			t.right.insert(key)
		}
	}
}

func (t *auxTree) find(key int) bool {
	if key == t.key {
		return true
	} else if key < t.key {
		if t.left == nil {
			return false
		}
		return t.left.find(key)
	} else {
		if t.right == nil {
			return false
		}
		return t.right.find(key)
	}
}

type tangoTree []*auxTree

func newTangoTree() tangoTree {
	return tangoTree{newAuxTree(0)}
}

func (t tangoTree) find(key int) bool {
	for _, aux := range t {
		if aux.find(key) {
			return true
		}
	}
	return false
}

func (t *tangoTree) update(key int) {
	leftAux, rightAux := t.split(key)
	*t = append(leftAux, append([]*auxTree{newAuxTree(key)}, rightAux...)...)
}

func (t tangoTree) split(key int) (left, right tangoTree) {
	leftAux, rightAux := new(auxTree), new(auxTree)
	for _, aux := range t {
		if aux.key < key {
			leftAux.insert(aux.key)
		} else {
			rightAux.insert(aux.key)
		}
	}
	left = tangoTree{leftAux}
	right = tangoTree{rightAux}
	return
}

File: dawn_cpp/tango.cpp
#include <iostream>
#include <vector>

template <typename T>
struct AuxTree {
    T key;
    AuxTree<T>* left = nullptr;
    AuxTree<T>* right = nullptr;

    AuxTree(T key) : key(key) {}

    ~AuxTree() {
        delete left;
        delete right;
    }

    void insert(T key) {
        if (key < this->key) {
            if (left == nullptr) {
                left = new AuxTree(key);
            } else {
                left->insert(key);
            }
        } else {
            if (right == nullptr) {
                right = new AuxTree(key);
            } else {
                right->insert(key);
            }
        }
    }

    bool find(T key) {
        if (key == this->key) {
            return true;
        } else if (key < this->key) {
            return (left != nullptr) && left->find(key);
        } else {
            return (right != nullptr) && right->find(key);
        }
    }
};

template <typename T>
struct TangoTree {
    std::vector<AuxTree<T>*> auxTrees;

    TangoTree() {
        auxTrees.push_back(new AuxTree<T>(0));
    }

    ~TangoTree() {
        for (auto aux : auxTrees) {
            delete aux;
        }
    }

    bool find(T key) {
        for (auto aux : auxTrees) {
            if (aux->find(key)) {
                return true;
            }
        }
        return false;
    }

    void update(T key) {
        std::vector<AuxTree<T>*> left, right;
        for (auto aux : auxTrees) {
            if (aux->key < key) {
                left.push_back(aux);
            } else {
                right.push_back(aux);
            }
        }
        auxTrees.clear();
        auxTrees.insert(auxTrees.end(), left.begin(), left.end());
        auxTrees.push_back(new AuxTree<T>(key));
        auxTrees.insert(auxTrees.end(), right.begin(), right.end());
    }
};